Gives The AI Hotkeys For Disconnecting From and Reconnecting To A Cyborg Shell

Jackraxxus opened the pull request:
## About The Pull Request
<img src="https://i.ibb.co/0cM6NHQ/Hotkey-PRImage2.png">
What it says on the tin.  Gives the AI hotkeys to connect and disconnect from its shell. I'm open to suggestions for better default hotkeys, I made sure to choose out-of-the-way keys with no overlaps because I have a vague memory about code-side keybind conflicts leading to a threatened forkening and I don't want that.

Also this is my first PR with more than 2 lines changed so I would appreciate a thorough and critical review of my code.
## Why It's Good For The Game

Since there are so many AI player admins (Who are all universally gigachads btw) it would be convenient to them to have an aghost-hotkey analogue for the AI. This change was specifically requested of me by one of these admins in fact.

I'm sure non-admin AIs would appreciate this convenience as well though.
## Changelog
:cl:
qol: The AI now has hotkeys for its shells, make sure to bind them to your preferred keys in your keybindings menu, under game preferences.
/:cl:

Jackraxxus committed:
Gives AI Hotkeys For Interacting With Its Shell

Tested and working on this commit

Jackraxxus committed:
makes things neater

I tested it again and this commit didn't break anything

Jackraxxus committed:
Some minor things annoyed me

Tested again and still works

Jackraxxus commented:
I made sure to update the PR's image in-line with the changes in that commit

ZephyrTFA commented on code/datums/keybinding/artificial_intelligence.dm:
```diff
@@ -0,0 +1,21 @@
+	weight = WEIGHT_AI
+
+/datum/keybinding/artificial_intelligence/can_use(client/user)
+	return isAI(user.mob)
```
This will cause issues because when in a shell the user is no longer typed to be the AI.
probably want to see if the user is a borg first and check for shell being true

Jackraxxus replied:
My original intent was that /artificial_intelligence hotkeys wouldn't be used while inside a shell (Right now for instance, there'd be no use to a deploy to shell hotkey while inside a shell). I'd have to see if it's compatible with can_use but I could probably implement a IsAIShell helper and put that as an OR in the return? I'd appreciate it if you could elaborate on these issues that you forsee, it might help me solve this is a neater way.

This could probably help address your last review as well, and get the disconnect hotkeys moved to this file in the process, but I'd have to check like !isAI(user) at the start of down(client/user) probably and I feel like that would be messy.

ZephyrTFA commented on code/datums/keybinding/artificial_intelligence.dm:
```diff
@@ -0,0 +1,21 @@
+	if(.)
+		return
+	var/mob/living/silicon/ai/our_ai = user.mob
+	our_ai.deploy_to_shell(our_ai.redeploy_action.last_used_shell)
```
What happens when this is null or the shell was destroyed?

Jackraxxus replied:
This is a good question, I'll add a check to make sure last_used_shell isn't null before it is passed through to deploy_to_shell.

Jackraxxus replied:
deploy_to_shell handles null / destroyed / dead shells by printing to chat "No usable AI shell beacons detected." Which is more feedback than my early return provided, so it's probably better to let it handle itself.

ZephyrTFA commented on code/datums/keybinding/robot.dm:
```diff
@@ -64,3 +64,22 @@
+	. = ..()
+	if(.)
+		return
+	var/mob/living/silicon/robot/shell/our_shell = user.mob
```
This can be used while not actually inside of a shell

Jackraxxus replied:
That's what the check in the next couple lines are for, if that's not good enough as stated above I could implement an IsAIShell() check and test that right at the start of down()? When I tested this as a cyborg on local it didn't spit any runtimes at me.

ZephyrTFA replied:
Reading over this again, this is fine and should work.

carshalash commented:
Why the fuck were you up at 4am making this PR?

Jackraxxus committed:
Addresses concerns in the second review comment

Tested and still works

Jackraxxus committed:
I changed my mind

deploy_to_shell handles nulls already I don't actually need this check. And it does it better than this check does. It's a strong, independent function who don't need no prior checking for nulls.
